### DISCUSSION TITLE
Performance Problem for Sentence Embeddings

### DISCUSSION THREAD

**Discussion by ayylemao** [pykeio/ort#365] [Q&A] [answered]
Hello,
I'm trying to port a classification engine from python to rust for easier and more portable deployment.
The whole thing will have to run on CPU only, but I'm having big trouble getting comparable time performance vs. Python.
The input data are about 5000 sentences in a .csv file.

Here the python snippet I'm using for comparison:
```python
import torch
from sentence_transformers import SentenceTransformer
import pandas as pd

torch.set_num_threads(16)

sentences = pd.read_csv("test_params.csv").params.to_list()

model = SentenceTransformer("sentence-transformers/all-MiniLM-L6-v2", device='cpu')

print(model.encode(sentences))
```

It is outperforming the rust code (compiled with --release) significantly. (10s for python, 19s for rust).

It takes about double the execution time.
The MiniLM model has been converted to ONNX via:
```
optimum-cli export onnx --model sentence-transformers/all-MiniLM-L6-v2 all-MiniLM-L6-v2-onnx
```

Here my current rust code on the same data:

```rust
use csv::{self, Reader};
 
use ort::error::Error;
use ort::session::{builder::GraphOptimizationLevel, Session};
use ort::value::Tensor;
 
use std::path::Path;
use tokenizers::Tokenizer;
 
fn main() {
    let mut rdr = Reader::from_path("test_params.csv").unwrap();
 
    let inputs: Vec<String> = rdr
        .records()
        .filter_map(|r| r.ok()) 
        .filter_map(|record| record.get(1).map(|s| s.to_string()))
        .collect();
 
    ort::init()
        .with_name("sentence_embedding")
        .commit()
        .unwrap();
 
    let model = Session::builder()
        .unwrap()
        .with_optimization_level(GraphOptimizationLevel::Level3)
        .unwrap()
        .with_intra_threads(16)
        .unwrap()
        .commit_from_file("all-MiniLM-L6-v2-onnx/model.onnx")
        .unwrap();
 
    let tokenizer =
        Tokenizer::from_file(Path::new("all-MiniLM-L6-v2-onnx/tokenizer.json")).unwrap();
 
    let encodings = tokenizer
        .encode_batch(inputs.clone(), true)
        .map_err(|e| Error::new(e.to_string()))
        .unwrap();
    let padded_token_length = encodings[0].len();
    let ids: Vec<i64> = encodings
        .iter()
        .flat_map(|e| e.get_ids().iter().map(|i| *i as i64))
        .collect();
    let mask: Vec<i64> = encodings
        .iter()
        .flat_map(|e| e.get_attention_mask().iter().map(|i| *i as i64))
        .collect();
 
    let a_ids = Tensor::from_array(([inputs.len(), padded_token_length], &*ids)).unwrap();
    let a_mask = Tensor::from_array(([inputs.len(), padded_token_length], &*mask)).unwrap();
 
    let outputs = model.run(ort::inputs![a_ids, a_mask].unwrap()).unwrap();
     
    println!("{:?}", outputs["sentence_embedding"].try_extract_tensor::<f32>().unwrap());
   ```

Can anybody give me some pointers where I'm going wrong? I can hardly imagine that ort is so much slower than the python implementations of the same models...

**Comment by decahedron1** [comment]
SBERT internally batches computations on calls to `SentenceTransformer.encode()`, defaulting to a batch size of 32. In Rust, however, it seems you are embedding all 5,000 sentences in one go, which might explain the performance gap.

If batching doesn't help, you might have better luck trying other optimized CPU EPs like XNNPACK (included in prebuilt) or oneDNN (requires compilation). There's also tools like the [transformers optimizer](https://onnxruntime.ai/docs/performance/transformers-optimization.html) & [`onnxsim`](https://github.com/daquexian/onnx-simplifier) but I don't know how much they'll help, if at all.

**Reply by ayylemao** [reply]
So I've been digging into this a little deeper. I built the libonnxruntime.so with oneDNN support and it only helped marginally. I also implemented the batched execution as you suggested. The execution time per batch (batch_size=32) on my machine with rust-onr lies still at about **~110ms**. 

Now to get a real honest comparison, i started using the onnx backend in python as well. Python sentence-transformer takes about **~40ms per batch**.

```python
import time
import torch
from sentence_transformers import SentenceTransformer
import pandas as pd
import onnxruntime as ort

sentences = pd.read_csv("test_params.csv").dropna().params.to_list()

model = SentenceTransformer("sentence-transformers/all-MiniLM-L6-v2", device='cpu', backend='onnx', 
                            model_kwargs={
                                "file_name":"onnx/model_O3.onnx",
                                "provider":"CPUExecutionProvider",
                                })

#model.save_pretrained("optimized_model")
batch_size = 32
nbatches = len(sentences)/batch_size
start = time.time()*1000
a = model.encode(sentences, batch_size=batch_size)
end = time.time()*1000
print(f"{(end-start)/nbatches:.3f}")
```

This also allowed me to save the exact onnx model used by the python code here for comparing it with the very same in the rust setup.
What I am also able to do is extract the onnx session parameters from python:
```
enable_cpu_mem_arena -> True
enable_mem_pattern -> True
enable_mem_reuse -> True
enable_profiling -> False
execution_mode -> ExecutionMode.ORT_SEQUENTIAL
execution_order -> ExecutionOrder.DEFAULT
get_session_config_entry -> <bound method PyCapsule.get_session_config_entry of <onnxruntime.capi.onnxruntime_pybind11_state.SessionOptions object at 0x7eff05a22a30>>
graph_optimization_level -> GraphOptimizationLevel.ORT_ENABLE_ALL
inter_op_num_threads -> 0
intra_op_num_threads -> 0
log_severity_level -> -1
log_verbosity_level -> 0
logid -> 
optimized_model_filepath -> 
profile_file_prefix -> onnxruntime_profile_
register_custom_ops_library -> <bound method PyCapsule.register_custom_ops_library of <onnxruntime.capi.onnxruntime_pybind11_state.SessionOptions object at 0x7eff05a22a30>>
use_deterministic_compute -> False
```

I tried emulating this as best as possible, including only using the CPUExecutionProvider since this is what python is explicitly told to use.

With this rust code, we still cannot get anywhere close to the performance of the python module, even when using the same backend and the same model.onnx.

How can this be?
The rust code is a messy, but i specifically only timed the actual model.run() part.

```rust
use csv::{self, Reader};
use ort::error::Error;
use ort::execution_providers::{
    CPUExecutionProvider, OneDNNExecutionProvider, XNNPACKExecutionProvider,
};
use ort::session::{builder::GraphOptimizationLevel, Session};
use ort::value::Tensor;
use std::ops::Add;
use std::path::Path;
use std::time::{Duration, Instant};
use tokenizers::Tokenizer;

const BATCH_SIZE: usize = 32;

fn main() {
    let mut rdr = Reader::from_path("test_params.csv").unwrap();
    let dylibpath = "onnxruntime/build/Linux/Release/libonnxruntime.so.1.22.0";
    let inputs: Vec<String> = rdr
        .records()
        .filter_map(|r| r.ok())
        .filter_map(|record| record.get(1).map(|s| s.to_string()))
        .collect();

    ort::init_from(dylibpath)
        .with_execution_providers([CPUExecutionProvider::default().with_arena_allocator().build()])
        .with_name("sentence_embedding")
        .commit()
        .expect("Failed to start ORT init!");

    let model = Session::builder()
        .unwrap()
        .with_optimization_level(GraphOptimizationLevel::Level3)
        .unwrap()
        .with_intra_op_spinning(false) .unwrap()
        .with_parallel_execution(false).unwrap()
        .with_memory_pattern(true).unwrap()
        .commit_from_file(
            "optimized_model/onnx/model_O3.onnx",
        )
        .unwrap();

    let tokenizer = Tokenizer::from_file(Path::new(
        "optimized_model/tokenizer.json",
    ))
    .unwrap();

    let start = Instant::now();
    let mut total: usize = 0;
    let mut accumul: Duration = Duration::new(0, 0);
    // **Process inputs in mini-batches**
    for (index, batch) in inputs.chunks(BATCH_SIZE).enumerate() {
        let encodings = tokenizer
            .encode_batch(batch.to_vec(), true)
            .map_err(|e| Error::new(e.to_string()))
            .unwrap();

        let padded_token_length = encodings[0].len();

        let ids: Vec<i64> = encodings
            .iter()
            .flat_map(|e| e.get_ids().iter().map(|i| *i as i64))
            .collect();
        let mask: Vec<i64> = encodings
            .iter()
            .flat_map(|e| e.get_attention_mask().iter().map(|i| *i as i64))
            .collect();
        let token_type_ids: Vec<i64> = vec![0; batch.len() * padded_token_length];

        let a_ids = Tensor::from_array(([batch.len(), padded_token_length], &*ids)).unwrap();
        let a_mask = Tensor::from_array(([batch.len(), padded_token_length], &*mask)).unwrap();
        let a_token_type_ids =
            Tensor::from_array(([batch.len(), padded_token_length], &*token_type_ids)).unwrap();

        // Run inference
        let sub_time = Instant::now();
        let outputs = model.run(ort::inputs!["input_ids" => a_ids, "attention_mask" => a_mask, "token_type_ids" => a_token_type_ids].unwrap()).unwrap();
        accumul = accumul.add(sub_time.elapsed());
        total = index;
    }
    println!("Per batch: {:?}", accumul.div_f64(total as f64 + 1.0));
    println!("Total Execution Time: {:3?}", start.elapsed());
}
```

**TL;DR**:
I tried optimizing Rust's ONNX inference to match Python's sentence-transformers, but Rust is still significantly slower (~110ms vs. ~40ms per batch).

- Built libonnxruntime.so with oneDNN – only marginal improvement.

- Implemented batched execution (batch_size=32).

- Ensured a fair comparison by using the same ONNX model and CPUExecutionProvider in both setups.

- Extracted Python's ONNX session parameters and mirrored them in Rust (to the best of my abbility).

Despite all this, Rust’s inference is still much slower. The timing is only for model.run(), so what’s causing this huge gap? 

Are the rust bindings to libonnxruntime.so just slower or is the onnx lib that python uses `onnxruntime_pybind11_state.cpython-310-x86_64-linux-gnu.so` just "built different"?

**Comment by ayylemao** [comment, marked as answer]
Ok after doing all this I have finally found the discrepancy. It had nothing to do with with the onnxruntime or the execution providers. 
It was all about the tokenizer.
The rust tokenizer package always pads the input_tokens up to 128 as stated in the tokenizer.config file.
Doesn't matter how long the actual sentence is.

The python implementation of sentence_transformers pads the input_ids dynamically to the max token_sequence within the batch.

If we do a fair comparison and feed the python code sentences that are 128 tokens in length (same as in the rust code), we get following performance per batch:

Python: ~130ms per batch
Rust: ~110ms per batch

So yeah 😊 

I guess I need to look into the tokenizer and how to get it to use a different padding strategy.
Thanks to @decahedron1 for looking into it!
